packages/gatsby/src/schema/infer/inference-metadata.ts

```diff
@@ -333,38 +414,44 @@ const updateValueDescriptor = (
     typeof typeInfo.example !== `undefined` ? typeInfo.example : value
 }
 
-const mergeObjectKeys = (dpropsKeysA, dpropsKeysB) => {
+const mergeObjectKeys = (
+  dpropsKeysA: ITypeInfoObject["dprops"] = {},
+  dpropsKeysB: ITypeInfoObject["dprops"] = {}
+): string[] => {
   const dprops = Object.keys(dpropsKeysA)
   const otherProps = Object.keys(dpropsKeysB)
   return [...new Set(dprops.concat(otherProps))]
 }
 
-const descriptorsAreEqual = (descriptor, otherDescriptor) => {
+const descriptorsAreEqual = (
+  descriptor?: IValueDescriptor,
+  otherDescriptor?: IValueDescriptor
+): boolean => {
   const types = possibleTypes(descriptor)
   const otherTypes = possibleTypes(otherDescriptor)
 
-  const childDescriptorsAreEqual = type => {
+  const childDescriptorsAreEqual = (type: string): boolean => {
     switch (type) {
       case `array`:
         return descriptorsAreEqual(
-          descriptor.array.item,
-          otherDescriptor.array.item
+          descriptor?.array?.item,
+          otherDescriptor?.array?.item
         )
       case `object`: {
         const dpropsKeys = mergeObjectKeys(
-          descriptor.object.dprops,
-          otherDescriptor.object.dprops
+          descriptor?.object?.dprops,
+          otherDescriptor?.object?.dprops
         )
         return dpropsKeys.every(prop =>
           descriptorsAreEqual(
-            descriptor.object.dprops[prop],
-            otherDescriptor.object.dprops[prop]
+            descriptor?.object?.dprops[prop],
+            otherDescriptor?.object?.dprops[prop]
           )
         )
       }
       case `relatedNode`:
       case `relatedNodeList`: {
-        return isEqual(descriptor.nodes, otherDescriptor.nodes)
+        return isEqual(descriptor?.nodes, otherDescriptor?.nodes)
```

Comment:
Looks like we've found a bug with a help of TS 😃 This should be something like:

```suggestion
      case `relatedNode`:
        return isEqual(
          descriptor?.relatedNode?.nodes,
          otherDescriptor?.relatedNode?.nodes
        )

      case `relatedNodeList`: {
        return isEqual(
          descriptor?.relatedNodeList?.nodes,
          otherDescriptor?.relatedNodeList?.nodes
        )
      }
```

Reply:
This one's going to take me a little longer as we get failing tests all over the place after making this change and the related one to the type description.

Reply:
Oh, yeah, the fix is probably not that straightforward. I think we can move it outside of the scope of this PR and handle it in a follow-up (as this is really a bugfix not directly related to TS conversion).

So I suggest to keep the old code but add ts-ignore to it with a short FIXME comment. I'll prepare a separate PR with a fix. Don't want to block this.